### PR TITLE
Support loading a model from `runs:/<run_id>/<artifact_path>`

### PR DESCRIPTION
**Pull request by harupy**
<details><summary>&#x1F6E0 DevTools &#x1F6E0</summary>
<p>

[Open in GitHub Codespaces](https://codespaces.new/harupy/mlflow/pull/13607?quickstart=1)

#### Install mlflow from this PR

```
pip install git+https://github.com/mlflow/mlflow.git@refs/pull/13607/merge
```

#### Checkout with GitHub CLI

```
gh pr checkout 13607
```

</p>
</details>

### Related Issues/PRs

 #xxx

### What changes are proposed in this pull request?

Support loading a model from `runs:/<run)_id>/<artifact_path>`

### How is this PR tested?

- [ ] Existing unit/integration tests
- [ ] New unit/integration tests
- [ ] Manual tests

### Does this PR require documentation update?

- [ ] No. You can skip the rest of this section.
- [ ] Yes. I've updated:
  - [ ] Examples
  - [ ] API references
  - [ ] Instructions

#### Is this a user-facing change?

- [ ] No. You can skip the rest of this section.
- [ ] Yes. Give a description of this change to be included in the release notes for MLflow users.

#### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/deployments`: MLflow Deployments client APIs, server, and third-party Deployments integrations
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>

#### How should the PR be classified in the release notes? Choose one:

- [x] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

#### Should this PR be included in the next patch release?

`Yes` should be selected for bug fixes, documentation updates, and other small changes. `No` should be selected for new features and larger changes. If you're unsure about the release classification of this PR, leave this unchecked to let the maintainers decide.

<details>
<summary>What is a minor/patch release?</summary>

- Minor release: a release that increments the second part of the version number (e.g., 1.2.0 -> 1.3.0).
  Bug fixes, doc updates and new features usually go into minor releases.
- Patch release: a release that increments the third part of the version number (e.g., 1.2.0 -> 1.2.1).
  Bug fixes and doc updates usually go into patch releases.

</details>

- [ ] Yes (this PR will be cherry-picked and included in the next patch release)
- [x] No (this PR will be included in the next minor release)

### CODE REVIEW

File: examples/mlflow-3/load_model_from_run_uri.py
Comment (harupy) [inline examples/mlflow-3/load_model_from_run_uri.py:1]: This is a test

File: mlflow/store/artifact/runs_artifact_repo.py [hunk 1]
```diff
@@ -128,7 +131,27 @@ def download_artifacts(self, artifact_path, dst_path=None):
...
+                "Attempting to download from the model associated with the run."
+            )
+            run_id = RunsArtifactRepository.parse_runs_uri(self.artifact_uri)[0]
+            client = mlflow.tracking.MlflowClient()
+            run = client.get_run(run_id)
+            # Search for models associated with the run
+            models = client.search_logged_models(
+                experiment_ids=[run.info.experiment_id],
+                filter_string=f"run_id = {run_id}",
+            )
```
Comment (harupy) [inline mlflow/store/artifact/runs_artifact_repo.py:148]: I think this part needs to be fixed.

Comment (harupy) [inline mlflow/store/artifact/runs_artifact_repo.py:148]:
```suggestion
            models = client.search_logged_models(
                experiment_ids=[run.info.experiment_id],
                filter_string=f"<expression that matches run_id and artifact_path?>",
                max_results=2,
            )
```

File: mlflow/store/artifact/runs_artifact_repo.py [hunk 2]
```diff
@@ -128,7 +131,27 @@ def download_artifacts(self, artifact_path, dst_path=None):
         Returns:
             Absolute path of the local filesystem location containing the desired artifacts.
         """
-        return self.repo.download_artifacts(artifact_path, dst_path)
+        try:
+            return self.repo.download_artifacts(artifact_path, dst_path)
+        except Exception:
+            _logger.debug(
+                "Failed to download artifacts. "
+                "Attempting to download from the model associated with the run."
```
Comment (dbczumar) [inline mlflow/store/artifact/runs_artifact_repo.py:139]:
```suggestion
                "Searching for logged models associated with the run instead."
```

File: mlflow/store/artifact/runs_artifact_repo.py (hunk 1 above)
Comment (dbczumar) [inline mlflow/store/artifact/runs_artifact_repo.py:148]: The Databricks backend doesn't actually support this yet. We should add it!

File: mlflow/store/artifact/runs_artifact_repo.py [hunk 3]
```diff
@@ -128,7 +131,27 @@ def download_artifacts(self, artifact_path, dst_path=None):
...
+            run_id = RunsArtifactRepository.parse_runs_uri(self.artifact_uri)[0]
+            client = mlflow.tracking.MlflowClient()
+            run = client.get_run(run_id)
+            # Search for models associated with the run
+            models = client.search_logged_models(
+                experiment_ids=[run.info.experiment_id],
+                filter_string=f"run_id = {run_id}",
+            )
+            if len(models) != 1:  # no model or multiple models found
+                raise
```
Comment (dbczumar) [inline mlflow/store/artifact/runs_artifact_repo.py:150]: There could totally be multiple logged models with the same source run ID. We shouldn't raise.

Instead, we should filter by name - the name should match the artifact path. If there are multiple models with the same name and source run ID, we should print a warning and pick the first one.

File: mlflow/store/artifact/runs_artifact_repo.py (hunk 1 above)
Comment (harupy) [reply mlflow/store/artifact/runs_artifact_repo.py:148]: Got it, I'll leave a TODO comment here.

File: mlflow/store/artifact/runs_artifact_repo.py (hunk 3 above)
Comment (harupy) [reply mlflow/store/artifact/runs_artifact_repo.py:150]: Sounds good!

File: mlflow/store/artifact/runs_artifact_repo.py [hunk 4]
```diff
@@ -128,7 +132,40 @@ def download_artifacts(self, artifact_path, dst_path=None):
...
+                    page_token=page_token,
+                )
+                for model in page:
+                    # Return the first model that matches the run_id and artifact_path
+                    if model.source_run_id == run_id and model.name == artifact_path:
+                        repo = get_artifact_repository(model.artifact_location)
+                        return repo.download_artifacts(
+                            artifact_path=".",  # root directory
+                            dst_path=dst_path,
+                        )
```
Comment (harupy) [inline mlflow/store/artifact/runs_artifact_repo.py:162]: We could search for models in all pages, but that's slow.

Comment (harupy) [inline mlflow/store/artifact/runs_artifact_repo.py:162]: The current implementation is:

- Just find the first match
- not sure if there are multiple matching models

**Review by dbczumar — APPROVED** [review]
LGTM!

**Pull request merged**